### ISSUE TITLE
Fix S2589: False Positive with foreach loop

### ISSUE THREAD

**Issue by nnpcYvIVl** [SonarSource/sonar-dotnet#2348]
The following code flags S2589 on the return line. Is this a false positive?

```csharp
using System.Collections.Generic;

namespace S2589_False_Positive
{
    static class Class
    {
        private static bool ListContainsAtLeastOnePositiveAndNoNegatives(List<int> list)
        {
            bool containspositive = false;
            bool containsnegative = false;

            foreach (int value in list)
            {
                if (value > 0)
                {
                    containspositive = true;
                }
                else if (value < 0)
                {
                    containsnegative = true;
                }
            }

            return containspositive && !containsnegative;
        }
    }
}
```

### Related information

* SonarC# Version: Latest
* Visual Studio Version: 15.9.9

**Comment by Corniel** [comment]
I agree. In case of `value == 0` the condition `containspositive && !containsnegative` is not true as rule S2589 claims. (expression always validate as true).

**Comment by Evangelink** [comment]
Hi there,

Thank you for the feedback! 

This is indeed a FP which is linked to our modelization of loops. I will keep this ticket open but I am afraid we won't be able to provide a fix in a short-term period.

**Comment by nnpcYvIVl** [comment]
@Evangelink Thanks for the confirmation and for details about the timeframe for a fix!

**Comment by andrei-epure-sonarsource** [comment]
Probably related to #1160
